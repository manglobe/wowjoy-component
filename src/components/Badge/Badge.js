import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const Wrap = styled.div`
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  color: rgba(0, 0, 0, 0.65);
  font-size: 14px;
  line-height: 1.5;
  list-style: none;
  position: relative;
  display: inline-block;
  color: unset;
  line-height: 1;
  ${props => props.defaultStyles};
`;

const statusColor = {
  success: "#52c41a",
  error: "#f5222d",
  processing: "#1890ff",
  warning: "#faad14",
  default: "#d9d9d9"
};

const StatusDot = styled.div`
  position: relative;
  top: -1px;
  display: inline-block;
  width: 6px;
  height: 6px;
  vertical-align: middle;
  border-radius: 50%;
  background-color: ${props => statusColor[props.status || "default"]};
`;

const StatusText = styled.span`
  margin-left: 8px;
  color: rgba(0, 0, 0, 0.65);
  font-size: 14px;
`;

const Count = styled.div`
  background-color: ${props => (props.color ? props.color : "#f5222d")};
  ${props =>
    props.dot
      ? `width: 6px;
  height: 6px;
  border-radius: 100%;`
      : `min-width: 20px;
  height: 20px;
  border-radius: 10px;
  padding: 0 6px;`}
  z-index: 10;
  color: #fff;
  font-weight: normal;
  font-size: 12px;
  line-height: 20px;
  white-space: nowrap;
  text-align: center;
  position: absolute;
  ${props =>
    props.offset
      ? `top:${props.offset[1]}px;right:-${props.offset[0]}px;`
      : `top: 0;right: 0;`}
  transform: translate(50%, -50%);
  transform-origin: 100% 0%;
  cursor: pointer;
`;
const BaseCount = ({ color, dot, offset, count, title, children }) => (
  <Count
    className="wjc-badge-count"
    color={color}
    dot={dot}
    offset={offset}
    count={count}
    title={title}
  >
    {children}
  </Count>
);
const StatusNode = ({ status, text }) => (
  <React.Fragment>
    <StatusDot className="badge-status-dot" status={status} />
    <StatusText className="badge-status-text">{text}</StatusText>
  </React.Fragment>
);
class Badge extends PureComponent {
  render() {
    const {
      className,
      defaultStyles,
      children,
      count,
      overflowCount,
      color,
      dot,
      offset,
      showZero,
      title,
      status,
      text
    } = this.props;
    const countNode = (showZero || count !== 0) && (
      <BaseCount {...{ color, dot, offset, count, title }}>
        {!dot && (count > overflowCount ? `${overflowCount}+` : count)}
      </BaseCount>
    );

    return (
      <Wrap
        defaultStyles={defaultStyles}
        className={`wjc-badge ${status ? ` wjc-status` : ""} ${className ||
          ""}`}
      >
        {!status && children}
        {status ? <StatusNode status={status} text={text} /> : countNode}
      </Wrap>
    );
  }
}

Badge.defaultProps = {
  overflowCount: 99,
  dot: false,
  showZero: false
};

Badge.propTypes = {
  className: PropTypes.string,
  defaultStyles: PropTypes.string,
  color: PropTypes.string,
  count: PropTypes.number,
  dot: PropTypes.bool,
  showZero: PropTypes.bool,
  offset: PropTypes.arrayOf(function(
    propValue,
    key,
    componentName,
    location,
    propFullName
  ) {
    if (key > 1) {
      return new Error(
        `offset should be an Array like [ number, number ], but length > 2`
      );
    }
    if (typeof propValue[key] !== "number") {
      return new Error(
        "Invalid prop `" +
          propFullName +
          "` supplied to" +
          " `" +
          componentName +
          "`. offset should be an Array like [ number, number ]."
      );
    }
  }),
  status: PropTypes.string,
  text: PropTypes.string,
  title: PropTypes.string
};
export default Badge;
