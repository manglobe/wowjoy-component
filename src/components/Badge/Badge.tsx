import * as React from "react";
import styled from "styled-components";
import { tuple } from "../_util/type";

const BadgeTypes = tuple(
  "success",
  "error",
  "processing",
  "warning",
  "default"
);
export type BadgeType = (typeof BadgeTypes)[number];

export interface StatusColor {
  [key: string]: string;
}
const statusColor: StatusColor = {
  success: "#52c41a",
  error: "#f5222d",
  processing: "#1890ff",
  warning: "#faad14",
  default: "#d9d9d9"
};

interface WrapProps {
  defaultStyles?: string;
}
interface StatusDotProps {
  status?: BadgeType;
}
interface StatusNodeProps extends StatusDotProps {
  text?: string;
}

interface BaseCountProps {
  color?: string;
  dot?: boolean;
  offset?: [number, number];
  count?: number;
  title?: string;
  children?: React.ReactNode;
}

export interface BadgeProps extends BaseCountProps, WrapProps, StatusNodeProps {
  className?: string;
  showZero?: boolean;
  overflowCount?: number;
}

const Wrap = styled.span<WrapProps>`
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  color: rgba(0, 0, 0, 0.65);
  font-size: 14px;
  line-height: 1.5;
  list-style: none;
  position: relative;
  display: inline-block;
  color: unset;
  line-height: 1;
  ${props => props.defaultStyles};
`;

const StatusDot = styled.span<StatusDotProps>`
  position: relative;
  top: -1px;
  display: inline-block;
  width: 6px;
  height: 6px;
  vertical-align: middle;
  border-radius: 50%;
  background-color: ${props => statusColor[props.status || "default"]};
`;

const StatusText = styled.span`
  margin-left: 8px;
  color: rgba(0, 0, 0, 0.65);
  font-size: 14px;
`;

const Count = styled.span<BaseCountProps>`
  background-color: ${props => (props.color ? props.color : "#f5222d")};
  ${props =>
    props.dot
      ? `width: 6px;
  height: 6px;
  border-radius: 100%;`
      : `min-width: 20px;
  height: 20px;
  border-radius: 10px;
  padding: 0 6px;`}
  z-index: 10;
  color: #fff;
  font-weight: normal;
  font-size: 12px;
  line-height: 20px;
  white-space: nowrap;
  text-align: center;
  position: absolute;
  ${props =>
    props.offset
      ? `top:${props.offset[1]}px;right:-${props.offset[0]}px;`
      : `top: 0;right: 0;`}
  transform: translate(50%, -50%);
  transform-origin: 100% 0%;
  cursor: pointer;
`;

const BaseCount = ({
  color,
  dot,
  offset,
  count,
  title,
  children
}: BaseCountProps) => (
  <Count
    className="wjc-badge-count"
    color={color}
    dot={dot}
    offset={offset}
    count={count}
    title={title}
  >
    {children}
  </Count>
);
const StatusNode = ({ status, text }: StatusNodeProps) => (
  <React.Fragment>
    <StatusDot className="badge-status-dot" status={status} />
    <StatusText className="badge-status-text">{text}</StatusText>
  </React.Fragment>
);

class Badge extends React.PureComponent<BadgeProps> {
  static defaultProps = {
    overflowCount: 99,
    dot: false,
    showZero: false
  };
  render() {
    const {
      className,
      defaultStyles,
      children,
      count,
      overflowCount,
      color,
      dot,
      offset,
      showZero,
      title,
      status,
      text
    } = this.props;
    const countNode = (showZero || count !== 0) && (
      <BaseCount {...{ color, dot, offset, count, title }}>
        {!dot && (count > overflowCount ? `${overflowCount}+` : count)}
      </BaseCount>
    );

    return (
      <Wrap
        defaultStyles={defaultStyles}
        className={`wjc-badge ${status ? `wjc-status ` : ""}${className || ""}`}
      >
        {!status && children}
        {status ? <StatusNode status={status} text={text} /> : countNode}
      </Wrap>
    );
  }
}

export default Badge;
