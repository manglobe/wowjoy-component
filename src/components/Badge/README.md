# Badge

Badge 组件

## 使用

-基础用法

```
import React from 'react';
import {Badge} from 'wowjoy-component';

const Foo=()=><Badge />
```

## APIs

|     属性      |                                  描述                                  |                          类型                           | 默认值 |
| :-----------: | :--------------------------------------------------------------------: | :-----------------------------------------------------: | :----: |
|   className   |                             顶层样式 class                             |                         string                          |        |
| defaultStyles |                              顶层默认样式                              |                         string                          |        |
|   children    |                                  内容                                  |                          node                           |        |
|     color     |                    自定义小圆点的颜色 string-3.16.0                    |
|     count     | 展示的数字，大于 overflowCount 时显示为\${overflowCount}+，为 0 时隐藏 |                         number                          |   -    |
|      dot      |                       不展示数字，只有一个小红点                       |                         boolean                         | false  |
|    offset     |                   设置状态点的位置偏移，格式为[x,y]                    |                     [number,number]                     |   -    |
| overflowCount |                            展示封顶的数字值                            |                         number                          |   99   |
|   showZero    |                     当数值为 0 时，是否展示 Badge                      |                         boolean                         | false  |
|    status     |                          设置 Badge 为状态点                           | Enum{'success','processing,'default','error','warning'} |   ''   |
|     text      |             在设置了 status 的前提下有效，设置状态点的文本             |                         string                          |   ''   |
| title [TODO]  |                    设置鼠标放在状态点上时显示的文字                    |                         string                          | count  |
