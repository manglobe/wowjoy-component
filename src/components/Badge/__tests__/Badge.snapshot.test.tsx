import * as React from "react";
import Badge from "../index";
import * as renderer from "react-test-renderer";
import * as TestUtils from "react-dom/test-utils";
import { MemoryRouter, Route } from "react-router-dom";
import { mount, configure } from "enzyme";
import * as ReactSixteenAdapter from "enzyme-adapter-react-16";
const adapter = ReactSixteenAdapter as any;
configure({ adapter: new adapter.default() });

const { renderIntoDocument } = TestUtils;

describe("Badge", () => {
  const reactObj = <Badge count={10} dot />;
  it("renders as dot correctly", () => {
    const tree = renderer.create(reactObj).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
