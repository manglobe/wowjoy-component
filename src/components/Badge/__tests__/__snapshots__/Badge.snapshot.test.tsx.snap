// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template renders correctly 1`] = `
<div
  className="wjc-badge   sc-bdVaJa VxzjQ"
>
  Template Content
  <div
    className="wjc-badge-count sc-bxivhb hqOToY"
  />
</div>
`;
